api-reference/lease/post.mdx

```diff
@@ -138,7 +138,7 @@ Returned if a single tenant can not be retrieved for any of your tenant records
 ```json Could not retrieve a single tenant
 {
    "error": {
-    "type": "multiple_landlord",
+    "type": "multiple_tenants_matched",
```

Comment:
I think the line below with the message could be clearer if it explained that multiple tenants were matched (1+ matches), rather than just not being able to match a single tenant which could mean 0 total matches as well

Reply:
Does this mean you're doing validation, within the API, before creating the lease?

Reply:
@thomaschristopherking 
So we have validation in the knowledge service already which raises
`raise DomainObjectNotFoundError(f"Could not find a unique tenant for creating the lease.")`
I'd be handling this in the API rather than doing the validation in the API.

@rohansampson 
Agreed, I'll modify the line.

Reply:
> @thomaschristopherking So we have validation in the knowledge service already which raises `raise DomainObjectNotFoundError(f"Could not find a unique tenant for creating the lease.")` I'd be handling this in the API rather than doing the validation in the API.

But the upsert operations are intended to be event-driven, and that's how the clients are implemented, so you wouldn't get that response. Were you intending on making direct REST requests?

Reply:
You're right, we wouldn't be able to the validation on the endpoints from the knowledge service unless we call them directly and not via the client.

For clients that use this endpoint in a RESTful manner (and not via future webhook integration) I think it makes sense to replicate this validation on the endpoint for instant feedback. Therefore I'll modify the task for this endpoint in notion to reflect that we must also add the validation on the endpoint. Does that sound sensible to you?

Reply:
The small problem is that validating and creating the record will not happen within a single transaction. So we could: determine the record is valid, but when we actually insert it the validation exception could still occur. The callee would have no idea this has failed.

Parallel question. For this specific exception, should the callee know about it? It's an internal error caused by how we distinguish between different avatars. I wouldn't expect our customers to know that we use contact details + a fuzzy name match. When someone provides contact details + first and last name, if there is a conflict then our data has become corrupt in some way. IMO, this is something we should resolve our end, because it indicates a bug, and then re-process the event.

So coming back to the original issue. I think the promise we should make is: we will do field-validation (e.g., we expect a float and you provided one). If the request passes that validation, then we will eventually upsert your record. We can make this promise because we decide whether to insert or update a record.

Is this feasible, or are there endpoints that genuinely require validation against the existing data? The only things I can think of are portfolio IDs, which the API knows about anyway.